### ISSUE TITLE
isEmpty/isNotEmpty for Collections

### ISSUE THREAD

**Issue by awhitford** [google/quiver-dart#512]
Similar to [quiver.strings](https://github.com/google/quiver-dart#quiverstrings)'s two useful functions:
```
/// Returns [true] if [s] is either null or empty.
bool isEmpty(String s) => s == null || s.isEmpty;

/// Returns [true] if [s] is a not empty string.
bool isNotEmpty(String s) => s != null && s.isNotEmpty;
```

Shouldn't there be equivalent methods for collections?  For example:
```
/// Returns [true] if [list] is either null or empty.
bool isListEmpty(List list) => list == null || list.isEmpty;

/// Returns [true] if [list] is a not empty.
bool isListNotEmpty(List list) => list != null && list.isNotEmpty;
```

**Comment by cbracken** [comment]
I agree it's a bit odd that we surface these for Strings but not for other types. 

In retrospect, I'm still not _entirely_ convinced that surfacing them was the right decision. These functions are mostly useful as quick drop-in utility functions in functional applications like `stringList.where(isNotEmpty)` which could equally easily be written `stringList.where((x) => x != null && x.isNotEmpty)`. Perhaps we should have named them `isNullOrEmpty` though that's trickier to negate.

We did discuss something along these lines years ago and ended up on the fence. This came up on the Java side as well in [Guava](https://github.com/google/guava) and [was rejected](https://stackoverflow.com/a/6921270), though I believe it ended up accidentally exposed in a helper class at some point and they may be stuck with it now :-) You can see Kevin Bourillion's StackOverflow reply above for rationale as to why it was rejected there.

I'm curious to hear a bit about some of the use-cases you can see for this. `Iterable<Iterable<T>>`s are a bit rarer than `Iterable<String>` in my experience.

**Comment by awhitford** [comment]
Should [extension-methods](https://dart.dev/guides/language/extension-methods) be used?  It's unclear on whether _null_ values are handled.

**Comment by cbracken** [comment]
So the answer there is that for the moment you _can_ write an extension method that works on null values, but that when Dart moves to non-null by default (NNDB), this will be disallowed. There's some detail in the [specification](https://github.com/dart-lang/language/blob/master/accepted/2.6/static-extension-members/feature-specification.md#semantics-of-invocations) for extension methods.

Here's a code snippet that demos this:
```
extension IsNullOrEmpty on List {
  bool get isNullOrEmpty {
    return this == null || this.isEmpty;
  }
}

int main() {
  List<int> list;
  if (list.isNullOrEmpty)
    print('empty');
}
```

**Comment by cbracken** [comment]
For what it's worth, I expect that once developers migrate to non-null by default, a lot of the need for these functions should go away -- same with Quiver's `Optional` type, which will be replaced by built-in language features for indicating whether nulls are permitted in a given context.
